tokio-test/src/io_stream.rs

```diff
@@ -0,0 +1,326 @@
+    /// call error.
+    pub fn write_error(&mut self, error: io::Error) -> &mut Self {
+        let error = Some(error.into());
+        self.tx.send(Action::WriteError(error)).unwrap();
+        self
+    }
+}
+
+impl Inner {
+    fn new(actions: VecDeque<Action>) -> (Inner, Handle) {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let rx = UnboundedReceiverStream::new(rx);
+
+        let inner = Inner {
+            actions,
+            sleep: None,
+            read_wait: None,
+            rx,
+            waiting: None,
+        };
+
+        let handle = Handle { tx };
+
+        (inner, handle)
+    }
+
+    fn poll_action(&mut self, cx: &mut task::Context<'_>) -> Poll<Option<Action>> {
+        Pin::new(&mut self.rx).poll_next(cx)
+    }
+
+    fn action(&mut self) -> Option<&mut Action> {
+        loop {
+            if self.actions.is_empty() {
+                return None;
+            }
+
+            match self.actions[0] {
+                Action::Read(ref mut data) => {
+                    if !data.is_empty() {
+                        break;
+                    }
+                }
+                Action::Write(ref mut data) => {
+                    if !data.is_empty() {
+                        break;
+                    }
+                }
+                Action::Next(ref mut data) => {
+                    if !data.is_empty() {
+                        break;
+                    }
+                }
+                Action::Wait(ref mut dur) => {
+                    if let Some(until) = self.waiting {
+                        let now = Instant::now();
+
+                        if now < until {
+                            break;
+                        }
+                    } else {
+                        self.waiting = Some(Instant::now() + *dur);
+                        break;
+                    }
+                }
+                Action::ReadError(ref mut error) | Action::WriteError(ref mut error) => {
+                    if error.is_some() {
+                        break;
+                    }
+                }
+            }
+
+            let _action = self.actions.pop_front();
+        }
+
+        self.actions.front_mut()
+    }
+}
+
+// ===== impl Inner =====
+
+
+impl Stream for Mock {
+    type Item = String;
+
+    fn poll_next(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+
+        loop {
+            if let Some(ref mut sleep) = self.inner.sleep {
+                ready!(Pin::new(sleep).poll(_cx));
+            }
+
+
+            // If a sleep is set, it has already fired
+            self.inner.sleep = None;
+            match ready!(self.inner.poll_action(_cx))  {
+                Some(action) => {
+                    self.inner.actions.push_back(action);
+                    continue;
+                }
+                None => {
+                    return Poll::Ready(None);
+                }
+            }
+        }
+
+    }
```

Comment:
The `poll_next` function never returns `Poll::Ready(Some(...))` anywhere, so the stream never produces any items.

Reply:
OMG!! I think I got it(copilot also helped me understand)
Guess the only thing that is missing is 
```if let Some(until) = self.waiting {
                        let now = Instant::now();

                        if now < until {
                            break;
                        }
                    } else {
                        self.waiting = Some(Instant::now() + *dur);
                        break;
                    }
```

for this method: 
```           Some(Action::Wait(dur)) => {
                    self.inner.sleep = Some(delay_for(dur));
                }
```

copilot could not help in those.

Reply:
The fact that you're trying to use `delay_for` shows that you're looking at old documentation. It was renamed to `sleep` in Tokio 1.x.y.

Please check out [the documentation for Sleep][1]. You need to use the `reset` method to update the deadline. There are examples in the docs.

[1]: https://docs.rs/tokio/latest/tokio/time/struct.Sleep.html